hetmech/bulk_computation.py

```diff
@@ -0,0 +1,87 @@
+import os
+
+import pandas
+import scipy.special
+import scipy.stats
+
+import hetmech.degree_group
+import hetmech.degree_weight
+import hetmech.hetmat
+
+
+def compute_save_dwpc(graph, metapath, damping=0.5, dense_threshold=1, dtype='float64', approx_ok=True,
+                      return_result=True):
+    """
+    Returns and saves a DWPC matrix. If already computed and return_result=True, will return saved result.
+    Automatically ignores inverses, corrupt files, and previously existing files.
+    """
+    metapath = graph.metagraph.get_metapath(metapath)
+    for inverse in (True, False):
+        if inverse:
+            mp = metapath.inverse
+        try:
+            return graph.read_path_counts(mp, 'dwpc', damping)
+        except FileNotFoundError:
+            continue
+        except:
+            # Catch all other file issues
+            os.remove(graph.get_path_counts_path(mp, 'dwpc', damping, 'sparse.npz'))
+    row, col, dwpc_matrix = hetmech.degree_weight.dwpc(graph, metapath, damping=damping,
+                                                       dense_threshold=dense_threshold, dtype=dtype)
+    path = graph.get_path_counts_path(metapath, 'dwpc', damping, None)
+    hetmech.hetmat.save_matrix(dwpc_matrix, path)
+    return row, col, dwpc_matrix
+
+
+def compute_save_dgp(hetmat, metapath, damping=0.5, compression='gzip', delete_intermediates=True):
+    """
+    Compute summary file of combined degree-grouped permutations (DGP). Aggregates across permutations,
+    deleting intermediates if delete_intermediates=True. Saves resulting files as compressed .tsv files
+    using compression method given by compression.
+    """
+    for mp in (metapath.inverse, metapath):
+        combined_path = hetmat.directory.joinpath(
+          'adjusted-path-counts', 'dwpc-0.5', 'degree-grouped-permutations', f'{mp}.tsv')
+        if combined_path.exists():
+            return
+
+    _, _, matrix = hetmat.read_path_counts(metapath, 'dwpc', damping)
+    matrix_mean = matrix.mean()
+
+    for name, permat in hetmat.permutations.items():
+        path = permat.directory.joinpath('degree-grouped-path-counts', 'dwpc-0.5', f'{metapath}.tsv')
```

Comment:
Let's discuss how to fix the w hardcode

Reply:
We should probably make some functions to receive these paths in the HetMat class?